### PR TITLE
partners: update pinecone client

### PR DESCRIPTION
**Pull request by aulorbe**
This PR updates the Pinecone client to `5.4.0`, as well as its dependencies (`pinecone-plugin-inference` and `pinecone-plugin-interface`). 

Note: `pinecone-client` is now simply called `pinecone`.

**Question for reviewer(s):** should this PR also update the `pinecone` dep in [the root dir's `poetry.lock` file](https://github.com/langchain-ai/langchain/blob/master/poetry.lock#L6729)? Was unsure. (I don't believe so b/c it seems pinned to a lower version likely based on 3rd-party deps (e.g. Unstructured).)

--
TW: @audrey_sage_

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208693659122374

### CODE REVIEW

File: libs/partners/pinecone/pyproject.toml [hunk 1]
```diff
@@ -21,15 +21,12 @@ disallow_untyped_defs = "True"
...
-[[tool.poetry.dependencies.numpy]]
-version = "^1"
-python = "<3.12"
-
-[[tool.poetry.dependencies.numpy]]
-version = "^1.26.0"
-python = ">=3.12"
+numpy = [
+    { version = "^1", python = "<3.12" },
+    { version = "^1.26.0", python = ">=3.12" }
```
Comment (aulorbe) [inline libs/partners/pinecone/pyproject.toml:28]: Just a structural change to make things cleaner

File: libs/partners/pinecone/langchain_pinecone/vectorstores.py [hunk 2]
```diff
@@ -74,6 +74,7 @@ class PineconeVectorStore(VectorStore):
                     dimension=1536,
                     metric="cosine",
                     spec=ServerlessSpec(cloud="aws", region="us-east-1"),
+                    deletion_protection="disabled",  # Defaults to "enabled"
```
Comment (jhamon) [inline libs/partners/pinecone/langchain_pinecone/vectorstores.py:77]: The default behavior is `deletion_protection="disabled"` So this comment is incorrect.

File: libs/partners/pinecone/pyproject.toml [hunk 3]
```diff
@@ -21,15 +21,12 @@ disallow_untyped_defs = "True"
 [tool.poetry.dependencies]
 python = ">=3.9,<3.13"
 langchain-core = "^0.3.15"
-pinecone-client = "^5.0.0"
+pinecone = ">=5.4.0"
```
Comment (jhamon) [inline libs/partners/pinecone/pyproject.toml:24]: This should be `"^5.4.0"`. `>=5.4.0` would include future breaking changes which may not be compatible.

Poetry docs have a nice explanation of how these various syntax elements in version numbers are interpreted. https://python-poetry.org/docs/dependency-specification/

**Review by jhamon — APPROVED** [review]
Thanks for taking this on, Audrey. There are a few small things I would address.

File: libs/partners/pinecone/pyproject.toml (hunk 3 above)
Comment (aulorbe) [reply libs/partners/pinecone/pyproject.toml:24]: Ah thank you so much!

File: libs/partners/pinecone/langchain_pinecone/vectorstores.py (hunk 2 above)
Comment (aulorbe) [reply libs/partners/pinecone/langchain_pinecone/vectorstores.py:77]: Eek, silly oversight. Thank you!

**Review by efriis — APPROVED** [review]

File: libs/partners/pinecone/langchain_pinecone/embeddings.py [hunk 4]
```diff
@@ -65,6 +65,19 @@ class PineconeEmbeddings(BaseModel, Embeddings):
...
 
+    @property
+    def async_client(self) -> aiohttp.ClientSession:
+        """Lazily initialize the async client."""
+        if self._async_client is None:
+            self._async_client = aiohttp.ClientSession(
+                headers={
+                    "Api-Key": self.pinecone_api_key.get_secret_value(),
+                    "Content-Type": "application/json",
+                    "X-Pinecone-API-Version": "2024-07",
```
Comment (jhamon) [inline libs/partners/pinecone/langchain_pinecone/embeddings.py:76]: You should probably update this API version header while you're here.

**Pull request merged**